core/src/triggers.rs

```diff
@@ -0,0 +1,176 @@
+            ))
+        } else {
+            self.hooks.insert(trigger.id.clone(), action).map_or_else(
+                || Ok(()),
+                |_| {
+                    Err(smartcontracts::Error::Repetition(
+                        InstructionType::Register,
+                        IdBox::TriggerId(trigger.id),
+                    ))
+                },
+            )
+        }
+    }
+
+    /// Remove a trigger from the [`TriggerSet`].
+    ///
+    /// # Errors
+    /// - If [`TriggerSet`] doesn't contain the trigger with the given [`EventFilter`].
+    /// Note that the [`EventFilter`] must be specified exactly.
+    pub fn remove(&self, id: trigger::Id) -> Result<(), smartcontracts::Error> {
+        self.hooks.remove(&id).map_or_else(
+            || {
+                Err(smartcontracts::Error::Repetition(
+                    InstructionType::Unregister,
+                    IdBox::TriggerId(id),
+                ))
+            },
+            |_| Ok(()),
+        )
+    }
+
+    /// Check if `self` contains `key`.
+    pub fn contains(&self, key: &trigger::Id) -> bool {
+        self.hooks.contains_key(key)
+    }
+
+    /// Add more repetitions to the hook identified by [`trigger::Id`].
+    ///
+    /// # Errors
+    /// - if trigger not found.
+    /// - if addition to remaining current trigger repeats
+    /// overflows. Indefinitely repeating triggers always cause an
+    /// overflow.
+    pub fn mint_repeats(&self, key: trigger::Id, amount: u32) -> Result<(), smartcontracts::Error> {
+        if self.hooks.contains_key(&key) {
+            let mut action = self.hooks.get_mut(&key).ok_or(FindError::Trigger(key))?;
+            let new_repeats = match action.value().repeats {
+                Repeats::Exactly(n) => n
+                    .checked_add(amount)
+                    .ok_or(smartcontracts::Error::Math(MathError::Overflow)),
+                _ => Err(smartcontracts::Error::Math(MathError::Overflow)),
+            }?;
+            action.value_mut().repeats = Repeats::Exactly(new_repeats);
+            Ok(())
+        } else {
+            Err(smartcontracts::Error::Find(Box::new(FindError::Trigger(
+                key,
+            ))))
+        }
+    }
+
+    /// Reduce repetitions of the hook identified by [`trigger::Id`].
+    ///
+    /// # Errors
+    /// - if trigger not found.
+    /// - if subtraction from remaining current trigger repeats
+    /// underflows. Indefinitely repeating triggers always cause an
+    /// underflow.
+    pub fn burn_repeats(&self, key: trigger::Id, amount: u32) -> Result<(), smartcontracts::Error> {
+        if self.hooks.contains_key(&key) {
+            let mut action = self.hooks.get_mut(&key).ok_or(FindError::Trigger(key))?;
+            let new_repeats = match action.value().repeats {
+                Repeats::Indefinitely => Err(smartcontracts::Error::Math(MathError::Overflow)),
+                Repeats::Exactly(n) => n
+                    .checked_sub(amount)
```

Comment:
this is the only difference between this method and `mint_repeats`. Maybe you could dedup?

Reply:
that's only a recommendation